corehq/util/pagination.py

```diff
@@ -55,14 +62,149 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
     while len_results:
         event_handler.page_start(total_emitted, *args, **kwargs)
         results = get_data_fn(*args, **kwargs)
-        start_time = datetime.datetime.utcnow()
+        start_time = datetime.utcnow()
         len_results = len(results)
 
         for item in results:
             yield item
 
         total_emitted += len_results
-        event_handler.page_end(total_emitted, datetime.datetime.utcnow() - start_time, *args, **kwargs)
+        event_handler.page_end(total_emitted, datetime.utcnow() - start_time, *args, **kwargs)
 
         if len_results:
             args, kwargs = next_args_fn(item, *args, **kwargs)
+
+
+class ResumableIteratorState(JsonObject):
+    doc_type = "ResumableIteratorState"
+    _id = StringProperty()
+    name = StringProperty()
+    timestamp = DateTimeProperty()
+    args = ListProperty()
+    kwargs = DictProperty()
+    retry = DictProperty()
+    progress = DictProperty()
+    complete = BooleanProperty(default=False)
+
+
+class ResumableFunctionIterator(object):
+    """Perform one-time resumable iteration over a data provider function.
+
+    Iteration can be efficiently stopped and resumed.
+
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `iteration_function` name to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param data_function: function to iterate over. Must return an list of data elements.
+    :param args_function: given the last data element of the previous ``data_function`` call and the previous
+    ``data_function`` args and kwargs return the next set of arguments for ``data_function``.
+    Initially called with ``None`` to get the fist set of arguments.
+    :param event_handler: Instance of ``PaginationEventHandler`` to be notified on page start and page end.
+    """
+
+    def __init__(self, iteration_key, data_function, args_function, event_handler=None):
+        self.iteration_key = iteration_key
+        self.data_function = data_function
+        self.args_function = args_function
+        self.event_handler = event_handler
+        self.iteration_name = '{}/{}'.format(iteration_key, data_function.__name__)
```

Comment:
What do you think of moving this out to `BlobDbBackendMigrator` and renaming `iteration_key` to `iteration_name` with a note in the docstring that `iteration_name` should be unique because it will be used to generate a unique identifier for persistent iteration state?

The reason I'm suggesting this is because the current way `ResumableDocsByTypeIterator` overwrites `data_function.__name__` seems like a bit of a hack, and obfuscates the real function's name, which could be useful in debugging, etc.